lib/Basic/FileSystem.cpp

```diff
@@ -14,12 +14,54 @@
 #include "llbuild/Basic/PlatformUtility.h"
 
 #include "llvm/ADT/STLExtras.h"
+#include "llvm/Support/FileSystem.h"
 #include "llvm/Support/Path.h"
 #include "llvm/Support/MemoryBuffer.h"
 
 #include <cassert>
 #include <cstring>
 
+// Cribbed from llvm, where it's been since removed.
+namespace {
+  using namespace std;
+  using namespace llvm;
+  using namespace llvm::sys::fs;
+
+  error_code _remove_all_r(StringRef path, file_type ft, uint32_t &count) {
+    if (ft == file_type::directory_file) {
+      error_code ec;
+      directory_iterator i(path, ec);
+      if (ec)
+        return ec;
+
+      for (directory_iterator e; i != e; i.increment(ec)) {
+        if (ec)
+          return ec;
+
+        file_status st;
+
+        if (error_code ec = i->status(st))
```

Comment:
This line doesn't seem safe to me. If my reading of the code is read, this uses `stat()` not `lstat()` and so this can end up recursing into a symlink which points at a directory, I think. Am I missing something?